store: alter deployment_schemas.version column to int

leoyvens opened the pull request:
The plan is to use this column to control migrations such as the one to a dedicated data source table described in #3405.

This changes that column to be an integer rather than an enum, so we don't need to add a DB enum variant on each new version.

lutter commented on store/postgres/migrations/2022-04-26-125552_alter_deployment_schemas_version/up.sql:
```diff
@@ -0,0 +1,62 @@
+drop view if exists info.all_sizes;
+drop materialized view if exists info.subgraph_sizes;
+
+alter table if exists deployment_schemas alter version type integer using 0;
```
I wouldn't say `if exists` here - if that table doesn't exist something is seriously wrong, and we should fail the migration.

lutter commented on store/postgres/src/primary.rs:
```diff
@@ -114,14 +114,30 @@ table! {
-    Relational,
+    /// Baseline version, in which:
+    /// - A relational schema is used.
+    /// - Each deployment has its own namespaces for entity tables.
```
typo: 'namespace'

lutter commented on store/postgres/src/primary.rs:
```diff
@@ -1116,7 +1131,13 @@ impl<'a> Connection<'a> {
+            shard,
+            src.deployment.clone(),
+            src.network.clone(),
+            src.schema_version,
```
This will be an important detail when we have more than one schema version: will copying create a deployment using the source's schema version or the latest version? My gut feeling is that it should always be the latest version, in which case `create_site` shouldn't take this as an argument and always use `CURRENT` for the new site.

lutter commented on store/postgres/src/primary.rs:
```diff
@@ -114,14 +114,30 @@ table! {
-/// with relational layout. Trying to do anything with a 'Split' subgraph
-/// will result in an error.
-#[derive(DbEnum, Debug, Clone, Copy)]
+/// This versions changes to the storage schema of deployments.
```
I don't understand this sentence

leoyvens commented:
Thanks for the review, I've addressed the comments and renamed `CURRENT` to `LATEST`, to better reflect that we intend to support multiple schema versions.

lutter commented on graph/src/components/store/err.rs:
```diff
@@ -50,6 +50,10 @@ pub enum StoreError {
+    #[error(
+        "found deployment schema version {0} which is not supported. Did you downgrade Graph Node?"
+    )]
+    UnsupportedDeploymentSchemaVersion(i32),
```
It might be friendlier if the error message was 
```
deployment '{Qm..}'({sgd123}) uses schema version {N} but this graph-node only supports versions up to {M}. Did you downgrade graph-node?
```

leoyvens replied:
I've improved this to show the max version, the deployment id will hopefully show up in the log or higher up in the error context. This required moving `DeploymentSchemaVersion` to the `graph` crate, but that was likely to be required eventually anyways.
